### PR TITLE
Make sure PNG memory is released in any case.

### PR DESCRIPTION
**Pull request by eustas**

### CODE REVIEW

File: lib/extras/dec/apng.cc [hunk 1]
```diff
@@ -405,6 +405,14 @@ Status DecodeImageAPNG(const Span<const uint8_t> bytes,
 
   std::vector<FrameInfo> frames;
 
+  auto before_return = [&](void*) {
+    // It is safe to invoke this method repeatedely; so make sure memory is
+    // released in any case.
+    png_destroy_read_struct(&png_ptr, &info_ptr, 0);
+  };
+  std::unique_ptr<void, decltype(before_return)> scope_guard((void*)1,
+                                                             before_return);
```
Comment (veluca93) [inline lib/extras/dec/apng.cc:414]: I'd rather

struct Guard {
 ~Guard() { ... }
} guard;

Comment (eustas) [reply lib/extras/dec/apng.cc:414]: Sure. It seems that there are 2 guards already -> time to extract template for that.

**Comment by eustas** [comment]
PTAL

File: lib/jxl/base/scope_guard.h [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+template <typename Callback>
+class ScopeGuard {
+ public:
+  // Discourage unnecessary moves / copies.
+  ScopeGuard(const ScopeGuard &) = delete;
+  ScopeGuard &operator=(const ScopeGuard &) = delete;
+  ScopeGuard &operator=(ScopeGuard &&) = delete;
+
+  // Some compilers have different RVO rules -> require move constructor.
```
Comment (veluca93) [inline lib/jxl/base/scope_guard.h:21]: More precisely, before C++17 RVO is not guaranteed to happen...

**Review by veluca93 — APPROVED** [review]

**Pull request merged**